subprojects/model-core/src/main/java/org/gradle/internal/reflect/annotations/impl/DefaultTypeAnnotationMetadataStore.java

```diff
@@ -349,6 +356,25 @@ private void processMethodAnnotations(Method method, Map<String, PropertyAnnotat
         }
     }
 
+    private void validateSetterForMutableType(PropertyAccessorType accessorType, Method method, ValidationErrorsBuilder errorsBuilder) {
+        Type setterType = accessorType.propertyTypeFor(method);
+        if (isRedundant(setterType)) {
+            String propertyName = accessorType.propertyNameFor(method);
+            String typeName = setterType.getTypeName();
+            errorsBuilder.recordError(String.format("'%s' of type '%s' has redundant setter method.", propertyName, typeName));
+        }
+    }
+
+    private boolean isRedundant(Type setter) {
+        if (!(setter instanceof Class<?>)) {
+            return false;
+        }
```

Comment:
This check should not be necessary if `propertyTypeFor()` returns `Class<?>`.

Reply:
What if it returns a class with generics or an array?